Add introducing ingress2gateway blog post (#43204)
kubernetes/website: content/en/blog/_posts/2023-10-18-gateway-api-time-to-switch-i2gw/index.md

Reviewer: If you wrap the source at say 100 characters, localization teams will thank you.

Author: > wrap the source at say 100 characters
you mean to wrap every line at 100?

Reviewer 2: Yeah +1, would make it easier to review. If you happen to use VSCode, I always recommend the [Rewrap extension](https://marketplace.visualstudio.com/items?itemName=stkb.rewrap) for this.

Author: so should I wrap at 100 or 80?

Reviewer 2: I usually do 80, but depends on repo, and I'd defer to @sftim's suggestion in this case since he's definitely got a lot more expertise here.

Author: wrapped at 100

---
layout: blog
title: "Time to Upgrade; Migrating to Gateway API"
date: 2023-10-19T10:00:00-08:00
slug: introducing-ingress2gateway
---

***Authors:*** Lior Lieberman (Google), Kobi Levi (independent)

Today we are releasing [ingress2gateway](https://github.com/kubernetes-sigs/ingress2gateway), a tool that can help you migrate from Ingress to Gateway API. Gateway API is just weeks away from graduating to GA, if you haven't upgraded yet, now's the time to think about it!

## Background

> In the ever-evolving world of Kubernetes, networking plays a pivotal role. As more applications are deployed in Kubernetes clusters, effective exposure of these services to clients outside the cluster becomes a critical concern. If you've been working with Kubernetes, you're likely familiar with the [Ingress API], which has been the go-to solution for managing external access to services.

Reviewer 2: Nit: Ingress and Gateway API are not limited to external access to Services.

Author: removed external from a few places. Resolving per our offline discussion

[Ingress API]:/docs/concepts/services-networking/ingress/

The Ingress API provides a way to route external traffic to your applications within the cluster, making it an indispensable tool for many Kubernetes users. Ingress has its limitations however, and as applications become more complex and the demands on your Kubernetes clusters increase, these limitations can become bottlenecks.

> Some of the limitations are;

Reviewer 2: ```suggestion
Some of the limitations are:
```

- **Insufficient common denominator** - by attempting to establish a common denominator for various HTTP proxies, Ingress can only accommodate basic HTTP routing, forcing more features of contemporary proxies like traffic splitting and header matching into provider-specific, non-transferable annotations.
- **Inadequate permission model** - Ingress spec configures both infrastructure and application configuration in one object. With Ingress, the cluster operator and application developer operate on the same Ingress object without being aware of each other’s roles. This creates an insufficient role-based access control and has high potential for setup errors. 
- **Lack of protocol diversity** - Ingress primarily focuses on HTTP(S) routing and does not provide native support for other protocols, such as TCP, UDP and gRPC. This limitation makes it less suitable for handling non-HTTP workloads.

> To overcome this, Gateway API is designed to provide a more flexible, extensible, and powerful way to manage external traffic to your services.

Reviewer 2: I think somewhere around here it would be useful to have another heading that made the following subheadings make more sense. Maybe something like this:

```suggestion
## Upgrading to Gateway API

To overcome this, Gateway API is designed to provide a more flexible, extensible, and powerful way to manage external traffic to your services.
```

Lastly, application developers configure route resources (such as HTTPRoutes), to manage configuration (e.g. timeouts, request matching/filter) and Service composition (e.g. path routing to backends)
Route resources define protocol-specific rules for mapping requests from a Gateway to Kubernetes Services.
HTTPRoute is for multiplexing HTTP or terminated HTTPS connections.
It's intended for use in cases where you want to inspect the HTTP stream and use HTTP request data for either routing or modification, for example using HTTP Headers for routing, or modifying them in-flight.

{{< figure src="gateway-api-resources.svg" caption="Diagram showing the APIs that make up Gateway and how they relate to each other" >}}

### Portability
With more than 20 [API implementations](https://gateway-api.sigs.k8s.io/implementations/#implementations), Gateway API is designed to be more portable across different implementations, clusters and environments. It helps reduce Ingress' reliance on non-portable, provider-specific annotations, making your configurations more consistent and easier to manage across multiple clusters.

> <!-- It is already supported on the 5 latest Kubernetes versions and the community commit to maintain support for the most recent 5 Kubernetes minor versions. -->

Reviewer 2: ```suggestion
Gateway API commits to supporting the 5 latest Kubernetes minor versions. That means that Gateway API currently supports Kubernetes 1.24+.
```

### Expressiveness
The Gateway API provides standard, Kubernetes-backed support for a wide range of features, such as header-based matching, traffic splitting, weight-based routing, request mirroring and more. With Ingress, these features need custom provider-specific annotations.

### Extensibility

> Gateway API is designed with extensibility as a core feature. Rather than enforcing a one-size-fits-all model, it offers the flexibility to link custom resources at multiple layers within the API's framework. This layered approach to customization ensures that users can tailor configurations to their specific needs without overwhelming the main structure. By doing so, the Gateway API facilitates more granular and context-sensitive adjustments, allowing for a fine-tuned balance between standardization and adaptability. This becomes particularly valuable in complex cloud-native environments where specific use cases require nuanced configurations.

Reviewer 2: Not a blocker, but something that may be worth being careful about here is that above you're saying that Ingress had too many implementation-specific annotations, and here you're showing that Gateway API can do roughly the same thing. Maybe worth clarifying that the key differences are that Gateway API has a much wider base set of features + a standard pattern for extensions that can be more expressive than annotations were on Ingress.

Reviewer 3: ```suggestion
### Extensibility
Gateway API is designed with extensibility as a core feature. Rather than enforcing a one-size-fits-all model, it offers the flexibility to link custom resources at multiple layers within the API's framework. This layered approach to customization ensures that users can tailor configurations to their specific needs without overwhelming the main structure. By doing so, the Gateway API facilitates more granular and context-sensitive adjustments, allowing for a fine-tuned balance between standardization and adaptability. This becomes particularly valuable in complex cloud-native environments where specific use cases require nuanced configurations. A critical difference that is worth mentioning is that Gateway API has a much broader base set of features and a standard pattern for extensions that can be more expressive than annotations were on Ingress.
```

How about that @robscott ?

Author: Thanks both, gonna incorporate this now.

## Upgrading to Gateway

> 

Reviewer 2: This feels like it needs some kind of transition instead of just jumping directly into the steps. Something like "If you've been convinced that it's time to migrate to Gateway API, here's how you can use Ingress2Gateway to do just that:"

Author: added something, let me know what you think

1. [Install a Gateway controller](https://gateway-api.sigs.k8s.io/guides/#installing-a-gateway-controller) OR [install the Gateway API CRDs manually](https://gateway-api.sigs.k8s.io/guides/#installing-gateway-api) 

2. install [ingress2gateway](https://github.com/kubernetes-sigs/ingress2gateway)

```

> go install github.com/kubernetes-sigs/ingress2gateway@latest

Reviewer 2: Assuming you'll change to something like this before publishing?
```suggestion
go install github.com/kubernetes-sigs/ingress2gateway@v0.1.0
```

Author: actually maybe we should use a stable tag?

Author: resolving as per our offline discussion

```
Provided you have a Go development environment installed, that command installs `ingress2gateway` to `$(go env GOPATH)/bin/ingress2gateway`.

> Alternatively, you can build it from source by following [Build from source](https://github.com/kubernetes-sigs/ingress2gateway/tree/95b4a0316bac8d28d72b871a210bbc9bd6191eb7#build-from-source)

Reviewer 2: I think I'd restructure this to read more like:

>If you have a Go development environment locally, you can install `ingress2gateway` with:
```
go install github.com/kubernetes-sigs/ingress2gateway@v0.1.0
```
>If not, you can download a binary at https://github.com/kubernetes-sigs/ingress2gateway/releases/tag/v0.1.0-rc1.

Presumably if they don't have a Go development environment, they're not going to be able to build this from source.

> 3. Once the tool is installed, you can use it to convert the ingress resources in your cluster to GatewayAPI resources.

Reviewer 2: ```suggestion
3. Once the tool is installed, you can use it to convert the ingress resources in your cluster to Gateway API resources.
```

```
ingress2gateway print
```

This above command will:

1. Read your Kube config file to extract the cluster credentials and the current active namespace.
1. Search for ingresses and provider-specific resources in that namespace.
1. Convert them to Gateway-API resources (Currently only Gateways and HTTPRoutes).
For other options refer run the tool with `-h` or refer to https://github.com/kubernetes-sigs/ingress2gateway#options

4. Review the converted Gateway API resources, validate them, and then apply them to your cluster.

> 5. Update your DNS to point to the new gateway. 

Reviewer 2: ```suggestion
5. Update your DNS to point to the new Gateway. 
```

> Note: make sure you test the newly created gateway before shifting the shifting 100% of the traffic.

Reviewer 2: I'd recommend adding this as a step between 4 and 5. Potentially with an example like `curl -h Host:example.com HTTP://{GATEWAY_ADDRESS}/foo/bar` that shows people how to get an address from Gateway status.

> 6. Make sure no traffic goes through your ingress and delete it.

Reviewer 2: ```suggestion
6. Once you've confirmed that no more traffic is going through your Ingress configuration, you can safely delete it.
```

## Wrapping up
Achieving reliable, scalable and extensible networking has always been a challenging objective.
The Gateway API is designed to improve the current Kubernetes networking standards like ingress and/or implementation specific annotations and CRDs.

It is a Kubernetes standard API, consistent across different platforms and implementations and most importantly it is future proof. The Gateway API is considered the Ingress’ successor, it has a sub team under SIG Network that actively work on it and manage the ecosystem and it is likely to receive more updates and community support.

> Ingress2gateway tool helps to ease the migration process for you, converting an existing Ingress object to equivalent objects from Gateway. To make migrations even easier, Ingress2gateway has recently added extensions to enable converting from provider-specific solutions, enabling it to translate implementation-specific CRDs or Ingress annotations to the closest Gateway API equivalent.

Reviewer 2: I think you've already made a good case for why people should upgrade to Gateway API to start this post, I'd recommend focusing on what is coming up next - any future plans for Ingress2Gateway, the upcoming v1.0 release of Gateway API, etc.

Author: made the changes, check "The Road Ahead" section

### Try it out and get involved
There are many resources to check out and learn more:

> - Get involved with the Kubernetes project on [GitHub](https://github.com/kubernetes-sigs/ingress2gateway)

Reviewer 2: ```suggestion
- Get involved with the Ingress2Gateway project on [GitHub](https://github.com/kubernetes-sigs/ingress2gateway)
```

> ---

Reviewer 2: Some high level non-blocking comments here. I think it might be better to restructure this into a rough outline like this:

**Intro**
Today we're releasing ingress2gateway, a tool that can help you migrate from Ingress to Gateway API. With Gateway API just weeks away from graduating to GA, it's time to consider upgrading if you haven't already...

**Why Upgrade?**
* New Features
* Role-Oriented Design
* Supported on 5 latest Kubernetes versions
* Already over 20 implementations of the API
* Can use the same APIs to configure mesh routing

**How to upgrade**
* Gateway API Getting Started guide for CRD + Controller installation
* Ingress2Gateway
* Run both in parallels, cutover DNS after manually verifying GW API routing

**Further Reading**

I think you already have most of the content for this, but IMO restructuring this post to follow this general layout would help it flow a bit better.

Author: > Supported on 5 latest Kubernetes versions

since what version?

Author: changed it to be closer to the suggested flow - PTAL

layout: blog
title: "TBD"
date: 2023-10-18T10:00:00-08:00
slug: <What is it?>
---

***Authors:*** Lior Lieberman (Google), Kobi Levi (independent)

> In the ever-evolving world of Kubernetes networking plays a pivotal role. As more applications are deployed in Kubernetes clusters, effective exposure of these services to clients outside the cluster becomes a critical concern. If you've been working with Kubernetes, you're likely familiar with the [Ingress API], which has been the go-to solution for managing external access to services.

Reviewer 2: ```suggestion
In the ever-evolving world of Kubernetes, networking plays a pivotal role. As more applications are deployed in Kubernetes clusters, effective exposure of these services to clients outside the cluster becomes a critical concern. If you've been working with Kubernetes, you're likely familiar with the [Ingress API], which has been the go-to solution for managing external access to services.
```

> [Ingress API]:https://kubernetes.io/docs/concepts/services-networking/ingress/

Reviewer: ```suggestion
[Ingress API]:/docs/concepts/services-networking/ingress/
```

The Ingress API provides a way to route external traffic to your applications within the cluster, making it an indispensable tool for many Kubernetes users. Ingress has its limitations however, and as applications become more complex and the demands on your Kubernetes clusters increase, these limitations can become bottlenecks.

Some of the limitations are;

> - **Insufficient common denominator** - by attempting to establish a common denominator for various HTTP proxies, Ingress can only accommodate basic HTTP routing, forcing more features of contemporary proxies like traffic splitting and header matching into provider-specific, non-transferable annotations.

Reviewer: ```suggestion
- **Insufficient common denominator** - by attempting to establish a common denominator for various HTTP proxies, Ingress can only accommodate basic HTTP routing, forcing implementations of Ingress to resort to custom annotations to deliver common features of contemporary proxies: traffic splitting. header matching, and more.
```

> - **Insufficient common denominator** - by attempting to establish a common denominator for various HTTP proxies, Ingress can only accommodate fundamental HTTP routing, forcing more features of contemporary proxies like traffic splitting, header matching and sticky sessions into provider-specific, non-transferable annotations.

Reviewer 2: ```suggestion
- **Insufficient common denominator** - by attempting to establish a common denominator for various HTTP proxies, Ingress can only accommodate basic HTTP routing, forcing more features of contemporary proxies like traffic splitting and header matching into provider-specific, non-transferable annotations.
```

Maybe just stick with features that GW API already supports today

Author: 😅

- **Inadequate permission model** - Ingress spec configures both infrastructure and application configuration in one object. With Ingress, the cluster operator and application developer operate on the same Ingress object without being aware of each other’s roles. This creates an insufficient role-based access control and has high potential for setup errors. 

> - **Lack of protocol diversity** - Ingress primarily focuses on HTTP(S) routing and does not provide native support for other protocols, such as TCP, UDP and GRPc. This limitation makes it less suitable for handling non-HTTP workloads.

Reviewer 2: ```suggestion
- **Lack of protocol diversity** - Ingress primarily focuses on HTTP(S) routing and does not provide native support for other protocols, such as TCP, UDP and gRPC. This limitation makes it less suitable for handling non-HTTP workloads.
```

> To overcome this, the Gateway API, ingress’ successor, is designed to provide a more flexible, extensible, and powerful way to manage external traffic to your services. It's not just a replacement for Ingress; it's a leap forward in Kubernetes networking.

Reviewer 2: Have to be careful here, Gateway isn't technically the replacement for Ingress since that API is not going away, this is just the next generation API which will get all new features.
```suggestion
To overcome this, Gateway API is designed to provide a more flexible, extensible, and powerful way to manage external traffic to your services.
```

## Introducing Gateway API

> Gateway API, with its official General Availability (GA) planned for late October, provides a standard Kubernetes API for ingress traffic control. It offers extended functionality, improved customization, and greater flexibility. By focusing on modular and expressive API resources, Gateway API makes it possible to describe a wider array of routing configurations and models.

Reviewer 2: ```suggestion
Gateway API is just weeks away from a GA release. It provides a standard Kubernetes API for ingress traffic control. It offers extended functionality, improved customization, and greater flexibility. By focusing on modular and expressive API resources, Gateway API makes it possible to describe a wider array of routing configurations and models.
```

The transition from Ingress API to Gateway API in Kubernetes is driven by advantages and advanced functionalities that the Gateway API offers, with its foundation built on four core principles: a role-oriented approach, portability, expressiveness and extensibility.

### A Role-Oriented Approach

Gateway API employs a role-oriented approach that aligns with the conventional roles within organizations involved in configuring Kubernetes service networking. This approach enables infrastructure engineers, cluster operators, and application developers to collectively address different aspects of the Gateway API.

For instance, infrastructure engineers play a pivotal role in deploying GatewayClasses, cluster-scoped resources that act as templates to explicitly define behavior for Gateways derived from them, laying the groundwork for robust service networking.

Subsequently, cluster operators utilize these GatewayClasses to deploy gateways.

> A Gateway in Kubernetes' Gateway-API defines how external traffic can be directed to Services within the cluster, essentially bridging non-Kubernetes sources to Kubernetes-aware destinations.

Reviewer: ```suggestion
A Gateway in Kubernetes' Gateway API defines how external traffic can be directed to Services within the cluster, essentially bridging non-Kubernetes sources to Kubernetes-aware destinations.
```

It represents a request for a load balancer configuration aligned with a GatewayClass’ specification.
The Gateway spec may not be exhaustive as some details can be supplied by the GatewayClass controller, ensuring portability.
Additionally, a Gateway can be linked to multiple Route references to channel specific traffic subsets to designated services.

Lastly, application developers configure route resources (such as HTTPRoutes), to manage configuration (e.g. timeouts, request matching/filter) and Service composition (e.g. path routing to backends)
Route resources define protocol-specific rules for mapping requests from a Gateway to Kubernetes Services.
HTTPRoute is for multiplexing HTTP or terminated HTTPS connections.
It's intended for use in cases where you want to inspect the HTTP stream and use HTTP request data for either routing or modification, for example using HTTP Headers for routing, or modifying them in-flight.

> ![The resources of Gateway API](gateway-api-resources.svg)

Reviewer: ```suggestion
![Diagram showing the APIs that make up Gateway and how they relate to each other](gateway-api-resources.svg)
```

Even better: use a [figure](https://gohugo.io/content-management/shortcodes/#figure) shortcode.

### Portability
Gateway API is designed to be more portable across different implementations, clusters and environments. It helps reduce Ingress' reliance on non-portable, provider-specific annotations, making your configurations more consistent and easier to manage across multiple clusters.

### Expressiveness

> The Gateway API sets Kuberentes standard support of a wide range of features, such as header-based matching, traffic splitting, weight-based routing, request mirroring and more. With Ingress, these features need custom provider-specific annotations.

Reviewer: “Kuberentes” is a typo.

```suggestion
The Gateway API provides standard, Kubernetes-backed support for a wide range of features, such as header-based matching, traffic splitting, weight-based routing, request mirroring and more. With Ingress, these features need custom provider-specific annotations.
```

### Extensibility

> Gateway API allows for custom resources to be linked at various layers of the API. This makes granular customization possible at the appropriate places within the API structure.

Reviewer 3: ```suggestion
Gateway API is designed with extensibility as a core feature. Rather than enforcing a one-size-fits-all model, it offers the flexibility to link custom resources at multiple layers within the API's framework. This layered approach to customization ensures that users can tailor configurations to their specific needs without overwhelming the main structure. By doing so, the Gateway API facilitates more granular and context-sensitive adjustments, allowing for a fine-tuned balance between standardization and adaptability. This becomes particularly valuable in complex cloud-native environments where specific use cases require nuanced configurations.
```

## Getting Started

### Gateway API Quick Start

The first thing you will need is a Kubernetes cluster. You can set up a new cluster using [KinD](https://kind.sigs.k8s.io/) with the following command:
```bash
kind create cluster --name introducing-i2gw
```

> When you create a service of type LoadBalancer in a Kubernetes cluster, typically, cloud providers (like AWS, GCP, Azure) automatically provision a cloud load balancer and allocate a public IP address for it. However, since KinD runs locally, there's no cloud provider available to automatically provision an external IP for such services.

Reviewer: ```suggestion
When you create a Service of `type: LoadBalancer` in a Kubernetes cluster, then typically cloud providers (like AWS, Azure or GCP) automatically provision a cloud load balancer and allocate a public IP address for it. However, since KinD runs locally, there's no cloud provider available to automatically provision an external IP address for such Services.
```

> This is where MetalLB comes in. MetalLB provides a network load-balancer implementation for Kubernetes clusters that do not have external load-balancers available, like KinD.

Reviewer: ```suggestion
This is where MetalLB can come in. MetalLB provides a network load-balancer implementation for Kubernetes clusters that do not have external load-balancers available; this could be KinD or it could be your bare-metal production cluster.
```

You'll need to [install](https://metallb.universe.tf/installation/) it in order to continue with this guide.

> Now you should have a working environment to start experimenting with Gateway-API. There are many great guides you can follow at https://gateway-api.sigs.k8s.io/guides/.

Reviewer: ```suggestion
Now you should have a working environment to start experimenting with Gateway
API. There are many great guides you can follow starting from [Getting started with Gateway API](https://gateway-api.sigs.k8s.io/guides/#getting-started-with-gateway-api).
```

> ### Migrating from Ingress to Gateway API using Ingress2Gateway

Reviewer: ```suggestion
### Migrating from Ingress to Gateway API using ingress2gateway
```
?

> Migrating from Ingress to the Gateway API may seem intimidating, but luckily Kubernetes SIG-Network provides a tool to simplify the process. Ingress2Gateway assists in the migration by converting your existing Ingress resources into Gateway API resources. To get started with ingress2gateway, you need to first install the tool.

Reviewer: ```suggestion
Migrating from Ingress to the Gateway API may seem intimidating, but luckily Kubernetes provides a tool to simplify the process. The `ingress2gateway` tool assists in the migration by converting your existing Ingress resources into Gateway API resources. To get started with `ingress2gateway`, you need to first install the tool:
```

If keeping the SIG name, it's “SIG Network” (no `-`).

Author: maybe a [ingress2gateway](https://github.com/kubernetes-sigs/ingress2gateway) instead of `ingress2gateway` ?

```
go install github.com/kubernetes-sigs/ingress2gateway@latest
```

> This will install ingress2gateway to $(go env GOPATH)/bin/ingress2gateway.

Reviewer: ```suggestion
Provided you have a Go development environment installed, that command installs `ingress2gateway` to `$(go env GOPATH)/bin/ingress2gateway`.
```

> Alternatively, you can build it from source using https://github.com/kubernetes-sigs/ingress2gateway#build-from-source.

Reviewer: ```suggestion
Alternatively, you can build it from source by following [Build from source](https://github.com/kubernetes-sigs/ingress2gateway/tree/d277aad7a2f55519f15a9bf99f5b733113702fda#build-from-source).
```

For a blog article, a permalink to a ref is better. Folks can update that later if they want to but at least it won't end up as a broken link.

Author: changed, thanks

Once the tool is installed, you can use it to convert the ingress resources in your cluster to GatewayAPI resources.

```
ingress2gateway print
```

This above command will:

1. Read your Kube config file to extract the cluster credentials and the current active namespace.
1. Search for ingresses and provider-specific resources in that namespace.
1. Convert them to Gateway-API resources (Currently only Gateways and HTTPRoutes).
For other options refer run the tool with `-h` or refer to https://github.com/kubernetes-sigs/ingress2gateway#options

> Review the converted Gateway API resources, validate them, and then apply them to your cluster.

Reviewer 2: I think ideally you'd want to pair this with some guidance such as how you'd use this with production workloads. 

Key points that are currently missing:

1. Follow the [Gateway API Getting Started Guide](https://gateway-api.sigs.k8s.io/guides/) to install the CRDs and an implementation of your choice.
2. Ingress and Gateway can run side-by-side, we recommend running both in parallel and only switching over DNS once you've been able to confirm that routing via Gateway API is working well.

Author: > Ingress and Gateway can run side-by-side, we recommend running both in parallel and only switching over DNS once you've been able to confirm that routing via Gateway API is working well.

We have spoken about it before I left for vacation. I think we should publish and official guidance in the gw official website, putting it here is not enough. Whats the go to solution we recommend for dns? I assuming it is [external-dns](https://github.com/kubernetes-sigs/external-dns)?

Asking so I can give references from here

Author: changed to reflect this. Resolving for now but open again if you need

## Wrapping up
Achieving reliable, scalable and extensible networking has always been a challenging objective.
The Gateway API is designed to improve the current Kubernetes networking standards like ingress and/or implementation specific annotations and CRDs.
By focusing on modular and expressive API resources, and a role-oriented approach, the Gateway API makes it possible to describe a wider array of routing configurations and models while improving security and ensuring smooth collaboration across different teams (infrastructure and application teams).

It is a Kubernetes standard API, consistent across different platforms and implementations and most importantly it is future proof. The Gateway API is considered the Ingress’ successor, it has a sub team under SIG-Network that actively work on it and manage the ecosystem and it is likely to receive more updates and community support.

> Ingress2gateway helps to ease the migration process for you, converting your existing Ingress resources to Gateway API resources. To make migrations _even easier_ Ingress2gateway has recently added extensions to enable converting from provider-specific solutions, enabling it to translate implementations specific CRDs or Ingress annotations to the relevant Gateway API configurations. 

Reviewer: I'd write:
```suggestion
The `ingress2gateway` tool helps to ease the migration process for you, converting an existing Ingress object to equivalent objects from Gateway . To make migrations even easier, `ingress2gateway` has recently added extensions to enable converting from provider-specific solutions, enabling it to translate some common implementation-specific custom resources or Ingress annotations to the closest Gateway API equivalent.
```

Author: thanks, I have changed it to more or less what you suggested

### Try it out and get involved
There are many resources to check out to learn more:
- Ingress2Gateway readme - https://github.com/kubernetes-sigs/ingress2gateway#ingress-to-gateway
- Gateway API Getting Started - https://gateway-api.sigs.k8s.io/guides/

> - Gateway API Implementations - https://gateway-api.sigs.k8s.io/implementations/#gateways

Reviewer: Please make the link text clickable for each of these.

---
layout: blog
title: "Time to Upgrade; Migrating to Gateway API"
date: 2023-10-24T10:00:00-08:00
slug: introducing-ingress2gateway
---

***Authors:*** Lior Lieberman (Google), Kobi Levi (independent)

Today we are releasing [ingress2gateway](https://github.com/kubernetes-sigs/ingress2gateway), a tool

> that can help you migrate from Ingress to Gateway API. Gateway API is just weeks away from

Reviewer: nit / could come in post merge:
I'd change _Ingress_ and _Gateway_ to hyperlinks, on this first mention, so people unfamiliar with these things can find out what they are

> graduating to GA, if you haven't upgraded yet, now's the time to think about it!

Reviewer: Ideally: explain the “GA” abbreviation

Author: like "GA (General Availability)" or did you think of something more than that?

Reviewer: “GA (General Availability)” is fine

  on the same Ingress object without being aware of each other’s roles. This creates an insufficient
  role-based access control and has high potential for setup errors. 
- **Lack of protocol diversity** - Ingress primarily focuses on HTTP(S) routing and does not provide
  native support for other protocols, such as TCP, UDP and gRPC. This limitation makes it less
  suitable for handling non-HTTP workloads.

## Gateway API

To overcome this, Gateway API is designed to provide a more flexible, extensible, and powerful way
to manage external traffic to your services.

Gateway API is just weeks away from a GA release. It provides a standard Kubernetes API for ingress
traffic control. It offers extended functionality, improved customization, and greater flexibility.
By focusing on modular and expressive API resources, Gateway API makes it possible to describe a
wider array of routing configurations and models.

The transition from Ingress API to Gateway API in Kubernetes is driven by advantages and advanced
functionalities that the Gateway API offers, with its foundation built on four core principles: a
role-oriented approach, portability, expressiveness and extensibility.

> ### A Role-Oriented Approach

Reviewer: ```suggestion
### A role-oriented approach
```

one-size-fits-all model, it offers the flexibility to link custom resources at multiple layers
within the API's framework. This layered approach to customization ensures that users can tailor
configurations to their specific needs without overwhelming the main structure. By doing so, the
Gateway API facilitates more granular and context-sensitive adjustments, allowing for a fine-tuned
balance between standardization and adaptability. This becomes particularly valuable in complex
cloud-native environments where specific use cases require nuanced configurations.

## Upgrading to Gateway

Migrating from Ingress to the Gateway API may seem intimidating, but luckily Kubernetes just
released a tool to simplify the process.
[ingress2gateway](https://github.com/kubernetes-sigs/ingress2gateway) assists in the migration by
converting your existing Ingress resources into Gateway API resources. Here is how you can get
started with Gateway API and using ingress2gateway:

1. [Install a Gateway
   controller](https://gateway-api.sigs.k8s.io/guides/#installing-a-gateway-controller) OR [install
   the Gateway API CRDs manually](https://gateway-api.sigs.k8s.io/guides/#installing-gateway-api) 

> 2. install [ingress2gateway](https://github.com/kubernetes-sigs/ingress2gateway) If you have a Go
> development environment locally, you can install `ingress2gateway` with:

Reviewer: ```suggestion
2. Install [ingress2gateway](https://github.com/kubernetes-sigs/ingress2gateway)
   If you have a Go development environment locally, you can install `ingress2gateway` with:
```

```
go install github.com/kubernetes-sigs/ingress2gateway@v0.1.0
```

This installs `ingress2gateway` to `$(go env GOPATH)/bin/ingress2gateway`.

Alternatively, follow the installation guide

> [here](https://github.com/kubernetes-sigs/ingress2gateway#installation).

Reviewer: Indent all of this by 3 spaces.

> 3. Once the tool is installed, you can use it to convert the ingress resources in your cluster to
>    Gateway API resources.

Reviewer: How about:
```suggestion
3. Once the tool is installed, you can use it to migrate from Ingress objects to equivalents from thr
   Gateway API subproject.
```

Author: Its not migrating, it is just a conversion, just initial conversion even.  The user need to check it, validate it and test it. 
When I read migration it has a slightly different meaning imo.

Author: resolving but feel free to open if needed

```
ingress2gateway print
```

This above command will:

> 1. Read your Kube config file to extract the cluster credentials and the current active namespace.

Reviewer: ```suggestion
1. Load your current Kubernetes client configuration including the active context, default namespace
   and authentication details.
```

> 1. Search for ingresses and provider-specific resources in that namespace.

Reviewer: ```suggestion
1. Search for Ingresses and provider-specific resources in that namespace.
```

1. Convert them to Gateway API resources (Currently only Gateways and HTTPRoutes). For other options

> refer run the tool with `-h` or refer to https://github.com/kubernetes-sigs/ingress2gateway#options

Reviewer: ```suggestion
  you can run the tool with `-h`, or refer to [https://github.com/kubernetes-sigs/ingress2gateway#options](https://github.com/kubernetes-sigs/ingress2gateway#options).
```

4. Review the converted Gateway API resources, validate them, and then apply them to your cluster.

5. Send test requests to your Gateway to check that it is working. You could get your gateway

>    address using `k get gateway <gateway-name> -n <namespace> -o

Reviewer: There's no `k` command as part of Kubernetes. Please clarify this.

Author: thanks for catching this!

   jsonpath='{.status.addresses}{"\n"}'`.

6. Update your DNS to point to the new Gateway. 

7. Once you've confirmed that no more traffic is going through your Ingress configuration, you can
   safely delete it.

## Wrapping up
Achieving reliable, scalable and extensible networking has always been a challenging objective. The

> Gateway API is designed to improve the current Kubernetes networking standards like ingress and/or

Reviewer: ```suggestion
Gateway API is designed to improve on current Kubernetes networking features such as Ingress,
and reduce the need for
```

Author: Thanks, accepted partially.

implementation specific annotations and CRDs.

It is a Kubernetes standard API, consistent across different platforms and implementations and most

> importantly it is future proof. The Gateway API is considered the Ingress’ successor, it has a sub

Reviewer: ```suggestion
importantly it is future proof. The Gateway API is considered the successor to Ingress; Gateway has a sub
```

essentially bridging non-Kubernetes sources to Kubernetes-aware destinations. It represents a
request for a load balancer configuration aligned with a GatewayClass’ specification. The Gateway
spec may not be exhaustive as some details can be supplied by the GatewayClass controller, ensuring
portability. Additionally, a Gateway can be linked to multiple Route references to channel specific
traffic subsets to designated services.

Lastly, application developers configure route resources (such as HTTPRoutes), to manage
configuration (e.g. timeouts, request matching/filter) and Service composition (e.g. path routing to
backends) Route resources define protocol-specific rules for mapping requests from a Gateway to
Kubernetes Services. HTTPRoute is for multiplexing HTTP or terminated HTTPS connections. It's
intended for use in cases where you want to inspect the HTTP stream and use HTTP request data for
either routing or modification, for example using HTTP Headers for routing, or modifying them
in-flight.

{{< figure src="gateway-api-resources.svg" caption="Diagram showing the APIs that make up Gateway

> and how they relate to each other" >}}

Reviewer: {{< figure src="gateway-api-resources.svg" alt="Diagram showing the APIs that make up Gateway and how they relate to each other. The APIs shown are GatewayClass, Gateway, and HTTPRoute; the Service API is also shown" caption="The APIs that make up Gateway" class="diagram-medium" >}}

---
layout: blog

> title: "Time to Upgrade; Migrating to Gateway API"

Reviewer: ```suggestion
title: "Use ingress2gateway to Help Migration to the Gateway API"
```

Let's not imply that everyone has to upgrade their cluster. Sometimes we want to make that assertion; this is not that time.

Reviewer 2: What about something like this? 

- "Introducing ingress2gateway, Simplifying Upgrades to Gateway API"

While I agree that we don't want to suggest that everyone needs to upgrade to Gateway API, I do think it's important to show that it is an upgrade.

Reviewer: ```suggestion
title: "Introducing ingress2gateway: Simplifying Upgrades to Adopt Gateway API
```

Author: I went with `Introducing ingress2gateway; Simplifying Upgrade to Gateway API`

Reviewer 2: IMO you either need "the Upgrade" or "Upgrades", would personally prefer "Upgrades" as it is more generic and would not be confused with a requirement to upgrade to Gateway API.

date: 2023-10-24T10:00:00-08:00
slug: introducing-ingress2gateway
---

***Authors:*** Lior Lieberman (Google), Kobi Levi (independent)

Today we are releasing [ingress2gateway](https://github.com/kubernetes-sigs/ingress2gateway), a tool
that can help you migrate from [Ingress](/docs/concepts/services-networking/ingress/) to [Gateway

> API](https://gateway-api.sigs.k8s.io). Gateway API is just weeks away from graduating to GA, if you

Reviewer: weeks? Or days?

Author: resolving this and leave the first comment open

limitations can become bottlenecks.

Some of the limitations are:

- **Insufficient common denominator** - by attempting to establish a common denominator for various
  HTTP proxies, Ingress can only accommodate basic HTTP routing, forcing more features of
  contemporary proxies like traffic splitting and header matching into provider-specific,
  non-transferable annotations.
- **Inadequate permission model** - Ingress spec configures both infrastructure and application
  configuration in one object. With Ingress, the cluster operator and application developer operate
  on the same Ingress object without being aware of each other’s roles. This creates an insufficient
  role-based access control and has high potential for setup errors. 
- **Lack of protocol diversity** - Ingress primarily focuses on HTTP(S) routing and does not provide
  native support for other protocols, such as TCP, UDP and gRPC. This limitation makes it less
  suitable for handling non-HTTP workloads.

## Gateway API

To overcome this, Gateway API is designed to provide a more flexible, extensible, and powerful way
to manage traffic to your services.

> Gateway API is just weeks away from a GA (General Availability) release. It provides a standard

Reviewer: Weeks? Or days?

Reviewer 2: Target is Oct 31, could slip to Nov 1. I think that's on the low end for "weeks" and high end for "days". Open to what you prefer here.

Kubernetes' Gateway API defines how external traffic can be directed to Services within the cluster,
essentially bridging non-Kubernetes sources to Kubernetes-aware destinations. It represents a
request for a load balancer configuration aligned with a GatewayClass’ specification. The Gateway
spec may not be exhaustive as some details can be supplied by the GatewayClass controller, ensuring
portability. Additionally, a Gateway can be linked to multiple Route references to channel specific
traffic subsets to designated services.

Lastly, application developers configure route resources (such as HTTPRoutes), to manage
configuration (e.g. timeouts, request matching/filter) and Service composition (e.g. path routing to
backends) Route resources define protocol-specific rules for mapping requests from a Gateway to
Kubernetes Services. HTTPRoute is for multiplexing HTTP or terminated HTTPS connections. It's
intended for use in cases where you want to inspect the HTTP stream and use HTTP request data for
either routing or modification, for example using HTTP Headers for routing, or modifying them
in-flight.

> {{< figure src="gateway-api-resources.svg" alt="Diagram showing the APIs that make up Gateway and how they relate to each other. The APIs shown are GatewayClass, Gateway, and HTTPRoute; the Service API is also shown" caption="The APIs that make up Gateway" class="diagram-medium" >}}

Reviewer 2: ```suggestion
{{< figure src="gateway-api-resources.svg" alt="Diagram showing the key resources that make up Gateway API and how they relate to each other. The resources shown are GatewayClass, Gateway, and HTTPRoute; the Service API is also shown" caption="The resources that make up Gateway API" class="diagram-medium" >}}
```
@sftim For some reason, the caption is not rendered any differently than the text in the [deploy preview](https://deploy-preview-43204--kubernetes-io-main-staging.netlify.app/blog/2023/10/25/introducing-ingress2gateway/#a-role-oriented-approach). Should we be doing something differently?
<img width="686" alt="Screenshot 2023-10-23 at 9 33 39 AM" src="https://github.com/kubernetes/website/assets/419322/726937ba-38eb-481c-8599-cc3610018f6f">

Author: I did not commit the new SVG, was waiting for consensus.
If we are all on board with the new one I will go ahead and do it. I saw one comment from you about preferring png

Reviewer 2: 🤦 I can be ok with SVG, but apparently I uploaded the wrong screenshot. I was trying to show that the caption looked out of place in the deploy preview:
<img width="594" alt="Screenshot 2023-10-24 at 9 35 08 AM" src="https://github.com/kubernetes/website/assets/419322/f108106f-068e-47a7-8ad8-2b9b698bbd5d">

Reviewer: Easy answer, then: omit the caption.

   3. Convert them to Gateway API resources (Currently only Gateways and HTTPRoutes). For other
   options you can  run the tool with `-h`, or refer to
   [https://github.com/kubernetes-sigs/ingress2gateway#options](https://github.com/kubernetes-sigs/ingress2gateway#options).

4. Review the converted Gateway API resources, validate them, and then apply them to your cluster.

5. Send test requests to your Gateway to check that it is working. You could get your gateway
   address using `kubectl get gateway <gateway-name> -n <namespace> -o
   jsonpath='{.status.addresses}{"\n"}'`.

6. Update your DNS to point to the new Gateway. 

7. Once you've confirmed that no more traffic is going through your Ingress configuration, you can
   safely delete it.

## Wrapping up

Achieving reliable, scalable and extensible networking has always been a challenging objective. The
Gateway API is designed to improve the current Kubernetes networking standards like ingress and
reduce the need for implementation specific annotations and CRDs.

It is a Kubernetes standard API, consistent across different platforms and implementations and most

> importantly it is future proof. The Gateway API is considered the successor to Ingress; Gateway API

Reviewer 2: A couple nits here:

1. Always prefer "Gateway API" over "The Gateway API" (comment applies throughout, but will avoid adding suggestion everywhere)
2. We can't say "successor" here as that can be construed as "replacement", but we are not getting rid of the Ingress API.
```suggestion
importantly it is future proof. Gateway API is the next generation of the Ingress API, but has a larger scope than that, expanding to tackle mesh and L4 routing as well. Gateway API
```

Reviewer: We don't have a good name for the family of API kinds related to Gateway, and this is a problem. How do docs writers separate Gateway the API kind from Gateway the family of APIs?

> Always prefer "Gateway API" over "The Gateway API" (comment applies throughout, but will avoid adding suggestion everywhere)

Can you PR that to https://kubernetes.io/docs/contribute/style/style-guide/ first? (I'm going to push back, though - using a definite article here feels more appropriate).

Reviewer: Also, avoid “L4”; try “layer 4” instead.

Reviewer 2: >How do docs writers separate Gateway the API kind from Gateway the family of APIs

Somehow I think we've generally managed to just refer to "Gateway" or "Gateway resource", for example https://gateway-api.sigs.k8s.io/api-types/gateway/. Naming is hard, but this project is first and foremost a set of APIs. For the record, the first name of the project was "Service APIs" and that was even worse.

>Can you PR that to https://kubernetes.io/docs/contribute/style/style-guide/ first? (I'm going to push back, though - using a definite article here feels more appropriate).

"Gateway API" is the name of the project. You wouldn't use "The Kubernetes". I could see using "The Gateway API" to refer to just the API for the Gateway resource, but it is decidedly not the name of the project. Unfortunately usage is currently mixed between "Gateway API" and "The Gateway API" (including in this blog post). If you feel strongly here, I can try to find some time to follow up with a PR to the style guide.

Reviewer: Mmm, OK. We like to get docs more polished for beta (and for in-tree things we usually get those nudges to happen).

I recommend that the first use of _Gateway API_ should appear in italics. That then fits with the style guide's advice.

Reviewer 2: Yeah, would be great to get Gateway API docs more polished. Definitely one of the many challenges of building an API out of tree. If there's anything sig-docs can help with on that front it would be very welcome.

> and Ingress2gateway are supported by a dedicated team under SIG Network that actively work on it and

Reviewer: ```suggestion
and `ingress2gateway` are supported by a dedicated team under SIG Network that actively work on it and
```

Author: We need to be consistent, its either I do ingress2gateway with lower "i" across all the blog, including the start of sentences. Or i do it with capital "I".
The repo name is with lower i but of course we can change that as well if we decide. 

Re code block, I do not want to put it in code block here. In this context it is "Gateway API and ingress2gateway are supported..." same as I don't put Gateway API in a code block I don't think I need one for ingress2gateway

Author: I changed the last two places to be with lower "i", do let me know if you prefer otherwise.

manage the ecosystem. It is also likely to receive more updates and community support.

### The Road Ahead

Ingress2gateway is just getting started. We're planning to onboard more providers, introduce support
for more types of Gateway API routes, and make sure everything syncs up smoothly with the ongoing
development of the Gateway API.

Excitingly, the Gateway API is also making significant strides. While v1.0 is about to launching,
there's still a lot of work ahead. This release incorporates many new experimental features, with
additional functionalities currently in the early stages of planning and development.

If you're interested in helping to contribute, we would love to have you! Please check out the
[community page](https://gateway-api.sigs.k8s.io/contributing/community/) which includes links to
the Slack channel and community meetings. We look forward to seeing you!!

### Useful Links

- Get involved with the Ingress2Gateway project on
  [GitHub](https://github.com/kubernetes-sigs/ingress2gateway)
- Open a new issue -

>   [Ingress2gateway](https://github.com/kubernetes-sigs/ingress2gateway/issues/new/choose), [Gateway

Reviewer: ```suggestion
  [ingress2gateway](https://github.com/kubernetes-sigs/ingress2gateway/issues/new/choose); [Gateway
```

---
layout: blog

> title: "TBD"

Reviewer 2: I think that's a great one, here's another idea: "Upgrading to Gateway API"

> title: "TBD"

Author: How about "Time to switch; Upgrading to Gateway API"

> title: "TBD"

Reviewer: Why _is_ it time to switch?

> title: "TBD"

Author: I went with "Time to Upgrade; Migrating to Gateway API"

Thoughts?

1. Convert them to Gateway-API resources (Currently only Gateways and HTTPRoutes).
For other options refer run the tool with `-h` or refer to https://github.com/kubernetes-sigs/ingress2gateway#options

Review the converted Gateway API resources, validate them, and then apply them to your cluster.

## Wrapping up
Achieving reliable, scalable and extensible networking has always been a challenging objective.
The Gateway API is designed to improve the current Kubernetes networking standards like ingress and/or implementation specific annotations and CRDs.
By focusing on modular and expressive API resources, and a role-oriented approach, the Gateway API makes it possible to describe a wider array of routing configurations and models while improving security and ensuring smooth collaboration across different teams (infrastructure and application teams).

It is a Kubernetes standard API, consistent across different platforms and implementations and most importantly it is future proof. The Gateway API is considered the Ingress’ successor, it has a sub team under SIG-Network that actively work on it and manage the ecosystem and it is likely to receive more updates and community support.

> To ease the migration process, you should definitely check Ingress2gateway. This tool will help you convert your ingresses to Gateway API resources. It has recently added support for extending it with provider-specific logic, meaning it can also translate implementations specific CRDs or ingress annotations to the relevant Gateway API configurations. 

Reviewer 4: ```suggestion
Ingress2gateway helps to ease the migration process for you, converting your existing Ingress resources to Gateway API resources. To make migrations _even easier_ Ingress2gateway has recently added extensions to enable converting from provider-specific solutions, enabling it to translate implementations specific CRDs or Ingress annotations to the relevant Gateway API configurations. 
```

- **Lack of protocol diversity** - Ingress primarily focuses on HTTP(S) routing and does not provide native support for other protocols, such as TCP, UDP and GRPc. This limitation makes it less suitable for handling non-HTTP workloads.

To overcome this, the Gateway API, ingress’ successor, is designed to provide a more flexible, extensible, and powerful way to manage external traffic to your services. It's not just a replacement for Ingress; it's a leap forward in Kubernetes networking.

## Introducing Gateway API
Gateway API, with its official General Availability (GA) planned for late October, provides a standard Kubernetes API for ingress traffic control. It offers extended functionality, improved customization, and greater flexibility. By focusing on modular and expressive API resources, Gateway API makes it possible to describe a wider array of routing configurations and models.

The transition from Ingress API to Gateway API in Kubernetes is driven by advantages and advanced functionalities that the Gateway API offers, with its foundation built on four core principles: a role-oriented approach, portability, expressiveness and extensibility.

### A Role-Oriented Approach

> The Gateway API employs a role-oriented approach that aligns with the conventional roles within organizations involved in configuring Kubernetes service networking. This approach enables infrastructure engineers, cluster operators, and application developers to collectively address different aspects of the Gateway API.

Reviewer 5: ```suggestion
Gateway API employs a role-oriented approach that aligns with the conventional roles within organizations involved in configuring Kubernetes service networking. This approach enables infrastructure engineers, cluster operators, and application developers to collectively address different aspects of the Gateway API.
```

We generally don't use a "the" before "Gateway API".

For instance, infrastructure engineers play a pivotal role in deploying GatewayClasses, cluster-scoped resources that act as templates to explicitly define behavior for Gateways derived from them, laying the groundwork for robust service networking.

<!-- TODO(liorlieberman) -->
<!-- Subsequently, cluster operators utilize these GatewayClasses to deploy gateways, brief whats gateways> -->

<!-- Lastly, application developers … -->

> ![The resources of the Gateway API](gateway-api-resources.png)

Reviewer 5: ```suggestion
![The resources of Gateway API](gateway-api-resources.png)
```
